### PR TITLE
Explain C# version defaults more clearly.

### LINKED ISSUE #15860: Is "compatibility" really the right term to use here?
The docs say:

&gt; the C# language may have features that rely on types or runtime components that are not available in every .NET implementation. This also ensures that for whatever target your project is built against, you get the highest compatible language version by default.

This suggests that C# 8 is incompatible with older target frameworks, in that code compiled successfully with C# 8 against those frameworks might not run.

However, as far as I can tell, C# 8 will give an error if you try to use the incompatible features (e. g. if you try to write a default implementation for an interface method). Therefore, I'm not convinced that C# 8 is truly incompatible vs. being "more restricted" when targeting older frameworks.

Is my interpretation correct? If so, I would suggest changing the wording. If not, I would suggest clarifying the nature of the incompatibility.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 5694e62b-c756-1f62-4014-2ceafa54c011
* Version Independent ID: 3f01c378-22e4-197c-a862-ab4a924506c7
* Content: [C# language versioning - C# Guide](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version)
* Content Source: [docs/csharp/language-reference/configure-language-version.md](https://github.com/dotnet/docs/blob/master/docs/csharp/language-reference/configure-language-version.md)
* Product: **dotnet-csharp**
* Technology: **csharp-language-reference**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### LINKED ISSUE #16335: "Why can't I select a different C# version?" not addressed
When trying to change the C# version on my project, in the Advanced Build Settings dialog box I found the language version drop-down greyed out, with a link saying "Why can't I select a different C# version?"  This link directed me to this page, which doesn't address this question at all.

---
#### Document details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 5694e62b-c756-1f62-4014-2ceafa54c011
* Version Independent ID: 3f01c378-22e4-197c-a862-ab4a924506c7
* Content: [C# language versioning - C# Guide](https://docs.microsoft.com/en-au/dotnet/csharp/language-reference/configure-language-version#feedback)
* Content Source: [docs/csharp/language-reference/configure-language-version.md](https://github.com/dotnet/docs/blob/master/docs/csharp/language-reference/configure-language-version.md)
* Product: **dotnet-csharp**
* Technology: **csharp-language-reference**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes [AB#1664867](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1664867)

Fixes #16335 
Fixes #15860

This article has several negative comments.

In this PR, I explain in more detail why these decisions were made. I also clarify why the UI was removed to select a different version, and point to how readers can edit the value to select an unsupported configuration

### CODE REVIEW

File: docs/csharp/language-reference/configure-language-version.md [hunk 1]
```diff
@@ -1,14 +1,19 @@
...
+The latest C# compiler determines a default language version based on your project's target framework or frameworks. Visual Studio doesn't provide a UI to change the value, but you can change it by editing the *csproj* file. The choice of default ensures that you use the latest language version compatible with your target framework. You benefit from access to the latest language features compatible with your project's target. This default choice also ensures you don't use a language that requires types or runtime behavior not available in your target framework. Choosing a language version newer than the default can cause hard to diagnose compile-time and runtime errors.
 
-The rules in this article apply to the compiler delivered with Visual Studio 2019 or the .NET Core 3.0 SDK. The C# compilers that are part of the Visual Studio 2017 installation or earlier .NET Core SDK versions target C# 7.0 by default. 
+These new defaults mean C# 8.0 is supported only on .NET Core 3.x and newer versions. Many of the newest features require library and runtime features introduced in .NET Core 3.x:
```
Comment (cartermp) [inline docs/csharp/language-reference/configure-language-version.md]:
```suggestion
These new defaults mean C# 8.0 (and higher) is supported only on .NET Core 3.x and newer versions. Many of the newest features require library and runtime features introduced in .NET Core 3.x:
```

Comment (cartermp) [reply docs/csharp/language-reference/configure-language-version.md]: Just so we don't have to go and update this every year.

File: docs/csharp/language-reference/configure-language-version.md [hunk 2]
```diff
@@ -1,14 +1,19 @@
...
+
+- Default interface member implementation requires new features in the .NET Core 3.0 CLR.
+- Async streams require the new types <xref:System.IAsyncDisposable?displayPropert=nameWithType>, <xref:System.IAsyncEnumerable%601?displayPropert=nameWithType>, and <xref:System.IAsyncEnumerator%601?displayPropert=nameWithType>.
+- Indexes and Ranges require the new types <xref:System.Index?displayProperty=nameWithType> and <xref:System.Range?displayProperty=nameWithType>.
+- Nullable reference types make use of several [attributes](../nullable-attributes.md) to provide better warnings. Those attributes were added in .NET 3.0, and .NET Framework libraries haven't been annotated with any of these attributes. That means nullable warnings may not accurately reflect potential issues.
```
Comment (cartermp) [inline docs/csharp/language-reference/configure-language-version.md]: Note - these are also unavailable in .NET Standard of any flavor, and Mono/xamarin. So it's only that usable for .NET Core.

**Review by cartermp — APPROVED** [review]
I think the overall changes are good, but some things to call out here.

**Comment by cartermp** [comment]
FWIW I'm not too confident that sentiments will improve here, since most people never touch the defaults, and those who do likely aren't going to be happy that the product has "moved the cheese".

File: docs/csharp/language-reference/configure-language-version.md (hunk 2 above)
Comment (BillWagner) [reply docs/csharp/language-reference/configure-language-version.md]: I reworded this slightly to clarify that only .NET Core 3.x has these applied.

**Review by cartermp — APPROVED** [review]

File: docs/csharp/language-reference/configure-language-version.md [hunk 3]
```diff
@@ -1,14 +1,19 @@
...
+The latest C# compiler determines a default language version based on your project's target framework or frameworks. Visual Studio doesn't provide a UI to change the value, but you can change it by editing the *csproj* file. The choice of default ensures that you use the latest language version compatible with your target framework. You benefit from access to the latest language features compatible with your project's target. This default choice also ensures you don't use a language that requires types or runtime behavior not available in your target framework. Choosing a language version newer than the default can cause hard to diagnose compile-time and runtime errors.
 
-The rules in this article apply to the compiler delivered with Visual Studio 2019 or the .NET Core 3.0 SDK. The C# compilers that are part of the Visual Studio 2017 installation or earlier .NET Core SDK versions target C# 7.0 by default. 
+These new defaults mean C# 8.0 (and higher) is supported only on .NET Core 3.x and newer versions. Many of the newest features require library and runtime features introduced in .NET Core 3.x:
```
Comment (tdykstra) [inline docs/csharp/language-reference/configure-language-version.md]: The choice of default doesn't determine what is supported, what is fully supported determines the choice of default.

```suggestion
These criteria mean that C# 8.0 (and higher) is the default language version only on .NET Core 3.x and newer versions. Many of the newest features require library and runtime features introduced in .NET Core 3.x:
```
Or omit "These criteria mean that":

> C# 8.0 (and higher) is the default language version only on .NET Core 3.x and newer versions. Many of the newest features require library and runtime features introduced in .NET Core 3.x:

**Review by tdykstra — APPROVED** [review]
Looks good, clearly explained.  One suggestion for you to consider.

File: docs/csharp/language-reference/configure-language-version.md (hunk 3 above)
Comment (BillWagner) [reply docs/csharp/language-reference/configure-language-version.md]: I omitted "these criteria..."

**Pull request merged**